File: haskell/src/Examples/Credibility.hs
-- | A Bayesian credibility model as a 'Learner'.
--
-- We use the Normal–Normal conjugate model: claim severity is assumed
-- @Normal(θ, σ²)@ for some unknown individual mean @θ@ with prior
-- @θ ~ Normal(μ₀, 1/κ₀)@. After each observation the posterior on @θ@
-- remains Normal with precision-weighted mean updates.
--
-- The classical Bühlmann credibility factor
--
-- > Z = n / (n + σ²/τ₀²)
--
-- falls out of the recursion: after @n@ observations the posterior
-- mean equals @Z * sample_mean + (1 - Z) * μ₀@.
module Examples.Credibility
    ( CredState (..)
    , credibilityLearner
    ) where

import Learner

data CredState = CredState
    { credMean :: !Double
    , credPrecision :: !Double
    }
    deriving (Show)

credibilityLearner ::
    -- | prior mean μ₀
    Double ->
    -- | prior precision on the mean κ₀ = 1/τ₀²
    Double ->
    -- | observation noise variance σ²
    Double ->
    Learner () Double
credibilityLearner mu0 kappa0 sigma2 =
    Learner
        (CredState mu0 kappa0)
        (\s () -> credMean s)
        ( \(CredState mu kappa) () x ->
            let pObs = 1 / sigma2
                kappa' = kappa + pObs
                mu' = (kappa * mu + pObs * x) / kappa'
             in CredState mu' kappa'
        )
        (\_ () _ -> ())

File: python/src/catins/learners/credibility.py
"""Bühlmann credibility learner.

This module provides a Bayesian credibility model as a Learner,
mirroring Examples.Credibility in the Haskell sketch.
"""

from dataclasses import dataclass

from catins.learner import Learner


@dataclass(frozen=True)
class CredState:
    """State for the credibility learner: (mean, precision)."""

    mean: float
    precision: float


class BuhlmannCredibility(Learner[CredState, None, float]):
    """Normal-Normal conjugate credibility model.

    # math: math.tex worked example §IV
    """

    def __init__(self, mu0: float, kappa0: float, sigma2: float):
        """Initialise with prior mean, prior precision, and observation variance.

        Args:
            mu0: Prior mean (μ₀).
            kappa0: Prior precision on the mean (κ₀ = 1/τ₀²).
            sigma2: Observation noise variance (σ²).
        """
        self.state = CredState(mu0, kappa0)
        self.sigma2 = sigma2

    def implement(self, _a: None) -> float:
        """Prediction is the current posterior mean."""
        return self.state.mean

    def update(self, _a: None, b: float) -> CredState:
        """Precision-weighted mean update.

        mu' = (kappa * mu + pObs * x) / (kappa + pObs)
        """
        p_obs = 1.0 / self.sigma2
        kappa_new = self.state.precision + p_obs
        mu_new = (self.state.precision * self.state.mean + p_obs * b) / kappa_new

        self.state = CredState(mu_new, kappa_new)
        return self.state

    def request(self, _a: None, _b: float) -> None:
        """Credibility is a terminal learner (empty request map)."""
        return
